ADR/0015-integration-service-two-phase-architecture.md

```diff
@@ -0,0 +1,109 @@
+# 15. The Two-phase Architecture of the Integration Service
+
+Date Documented: 2023-02-09
+
+## Status
+
+In review
+
+## Context
+
+The Integration Service is in charge of running integration test pipelines by executing
+the Tekton pipeline for each user-defined IntegrationTestScenario.
+The main payloads that are being tested are Snapshots which contain references to all
+Components that belong to the Application, along with their images.
+
+One problem faced by the integration service is caused by the fact that testing an
+application with Tekton pipelines takes time. During the course of testing an application,
+multiple component builds can happen in quick succession, leading to a potential race
+condition between different Snapshots that are created for each of those builds.
+This would primarily manifest itself by two Snapshots getting promoted in quick
+succession but neither of them having the latest images contained within them.
+
+## Decision
+
+In order to protect against the race conditions, the integration service will leverage a
+two phase approach to testing. The phases will consist of a Component phase that will
+always be executed and the optional Composite phase which would come into play only when
+race conditions between component builds are detected.
+
+### Component phase
+
+When a single component image is built, the Integration Service tests the application
+by creating a Snapshot. All Components with their images from the Global Candidate List
+are included within the Snapshot and then the Component that was newly built is
+updated/overwritten to complete the Snapshot creation.
+
+After all test pipelines for the Snapshot finish successfully, the Integration service
+updates the Global Candidate List with the newly built Component image and promotes
+the Snapshot. This is done only if the rest of the Global Candidate List matches
+the rest of the tested Snapshot components. If the Global Candidate List doesn't
+match, the Snapshot status is marked as invalid and the testing goes into the
+Composite phase.
+
+![](../diagrams/ADR-0015/component-phase.jpg)
+
+### Composite phase
+
+The Composite phase is used when the Global Candidate List changes while testing a
+Snapshot in the Component phase. The Integration service checks this before promoting
+a Snapshot at the end of the Component phase.
+
+The Composite phase exists to resolve a race condition when teams merge multiple PRs
+to multiple components of the same application at nearly the same time.
+When multiple components are built at the same time, the Integration Service tests
+the application by creating a composite Snapshot using multiple Components updated
+to use the newly built images.
+
+After the composite Snapshot passes testing, the entire Global Candidate List is updated
+according to its contents.
```

Comment:
I think that we have to contend with another race condition here. I opened https://issues.redhat.com/browse/HACBS-1779 to discuss it.

Reply:
@ralphbean based on your comments, should we drop this sentence entirely and modify the Integration service so it doesn't update the Global Candidate List at the end of the Composite phase?

Reply:
Yes, I think so.

Reply:
It is replaced with the sentence above during the _component_ phase:

> After all test pipelines for the Snapshot finish successfully, the Integration service updates the Global Candidate List with the newly built Component image

Reply:
Ack, removed the sentence and only mentioned promoting the Composite snapshot according to user preferences